File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
def solution(queue1, queue2):
    ns1,ns2 = {0:0},{0:0}
    
    hap1 = 0
    for i in range(len(queue1)):
        hap1+=queue1[i]
        ns1[hap1] = i+1
    
    hap2 = 0
    for i in range(len(queue2)):
        hap2+=queue2[i]
        ns2[hap2] = i+1
    
    hap=0
    for i in range(len(queue1)-1,-1,-1):
        hap-=queue1[i]
        ns2[hap] = len(queue2) + i
    hap=0
    for i in range(len(queue2)-1,-1,-1):
        hap-=queue2[i]
        ns1[hap] = len(queue1) + i
    
    
    if (hap1+hap2)%2:
        # print('case1')
        return -1
    
    result = len(queue1)*2 + len(queue2)*2+1
    cnt=0
    for key in ns1:
        key2 = (hap2-hap1)//2 + key
        # print('key',key,key2)
        if key2 in ns2:
            # print('possible', ns1[key], ns2[key2])
            result = min(result, ns1[key] + ns2[key2])
        cnt+=1
        if cnt == len(queue1):
            break
    cnt = 0
    for key in ns2:
        key2 = (hap1-hap2)//2 + key
        # print('key',key,key2)
        if key2 in ns1:
            # print('possible', ns2[key], ns1[key2])
            result = min(result, ns2[key] + ns1[key2])
        cnt+=1
        if cnt == len(queue1):
            break

    # print('result',result)
    if result ==len(queue1)*2 + len(queue2)*2+1:
        return -1
    
    return result

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

public class Solution {
    public static int solution(int[] queue1, int[] queue2) {
        Map<Long, Integer> ns1 = new LinkedHashMap<>();
        Map<Long, Integer> ns2 = new LinkedHashMap<>();
        ns1.put(0L, 0);
        ns2.put(0L, 0);

        long hap1 = 0;
        for (int i = 0; i < queue1.length; i++) {
            hap1 += queue1[i];
            ns1.put(hap1, i + 1);
        }

        long hap2 = 0;
        for (int i = 0; i < queue2.length; i++) {
            hap2 += queue2[i];
            ns2.put(hap2, i + 1);
        }

        long hap = 0;
        for (int i = queue1.length - 1; i >= 0; i--) {
            hap -= queue1[i];
            ns2.put(hap, queue2.length + i);
        }

        hap = 0;
        for (int i = queue2.length - 1; i >= 0; i--) {
            hap -= queue2[i];
            ns1.put(hap, queue1.length + i);
        }

        if ((hap1 + hap2) % 2 != 0) {
            return -1;
        }

        int maxResult = queue1.length * 2 + queue2.length * 2 + 1;
        int result = maxResult;

        int cnt = 0;
        for (Map.Entry<Long, Integer> entry : ns1.entrySet()) {
            long key = entry.getKey();
            long key2 = (hap2 - hap1) / 2 + key;
            if (ns2.containsKey(key2)) {
                result = Math.min(result, ns1.get(key) + ns2.get(key2));
            }
            cnt++;
            if (cnt == queue1.length) break;
        }

        cnt = 0;
        for (Map.Entry<Long, Integer> entry : ns2.entrySet()) {
            long key = entry.getKey();
            long key2 = (hap1 - hap2) / 2 + key;
            if (ns1.containsKey(key2)) {
                result = Math.min(result, ns2.get(key) + ns1.get(key2));
            }
            cnt++;
            if (cnt == queue2.length) break;
        }

        if (result == maxResult) {
            return -1;
        }

        return result;
    }
}
